### PR TITLE
chore: Refactor control reader Part 2

### PR DESCRIPTION
**Pull request by enikonovad**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary
A second part of refactoring of control_reader. This PR touches the process of reading Editable Content from the markdown.
- Refactored a good chunk of parts reading from control_reader
- Removed duplicated code and logic
- Now each node in the tree has a part object in its content (a part is defined by ## in the markdown file)
- Based on the section name a specific format for part_id and part_name will be picked

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/core/markdown/docs_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/docs_markdown_node.py:63]: I see this `__new__` has been in the code for some time - but it is the only time it is used.  Just want to confirm there is a reason this classmethod is special and needs it.

File: trestle/core/markdown/control_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/control_markdown_node.py:90]: Calling it `a_part` is odd to me.  Partly because a statement could have `Part a.` in it.  Is there a better name to describe what it is and what it contains?  Some doc text would help also.

File: trestle/core/markdown/base_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/base_markdown_node.py:152]: I think this function is dead and is not being called.  `level` isn't used and the function has no code coverage.  But it still may be needed in terms of the overall API...

File: trestle/core/markdown/control_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/control_markdown_node.py:57]: Please use caps for enums.  I think it's fairly standard convention and it is the convention in trestle - except in the auto-generated OSCAL python classes.  And it would be good to add doc strings on Part vs. part - e.g. is `## part foo` the same as `## Part foo`

File: trestle/common/const.py
Comment (fsuits) [inline trestle/common/const.py:232]: I think it's worth changing this to have `(?i)` in front to indicate directly that case is ignored when matching the string.  Then you can remove all the `lower()` references in the code that do matching.  Otherwise if a `lower()` is omitted it could be a bug.

File: trestle/core/markdown/control_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/control_markdown_node.py:204]: Should we have debug message or warning for undefined type?  It may represent content the user meant as an important part and it is being dropped - so I think a warning.

File: trestle/core/markdown/control_markdown_node.py
Comment (fsuits) [inline trestle/core/markdown/control_markdown_node.py:269]: Is there a way to provide more context for this error so the user knows where the problem is?  Perhaps the check needs to be before this call.

**Review by fsuits — APPROVED** [review]
LGTM after addressing comments - many of which are style and naming.  Before merge I think we should go over the terms and make sure they are used correctly: part, section and guidance.  So please hold until after discussion.

File: trestle/core/markdown/docs_markdown_node.py
Comment (enikonovad) [reply trestle/core/markdown/docs_markdown_node.py:63]: This is a fairly special use case for this class. This is a classmethod factory method to create the instance of MarkdownNode. The reason for this design choice was that each MarkdownNode instance operates on itself and builds the tree recursively. It needs to perform some subroutines (i.e. compute the max heading level, key, etc) and contain some information in each node (defined in __init__). Because we don't want to call `__init__` before we compute those this approach was used

File: trestle/core/markdown/control_markdown_node.py
Comment (enikonovad) [reply trestle/core/markdown/control_markdown_node.py:90]: `a_part` could represent any part from the markdown. A part is something that starts with two or more hashes. Let me think if there is any other generic enough name for this

File: trestle/core/markdown/base_markdown_node.py
Comment (enikonovad) [reply trestle/core/markdown/base_markdown_node.py:152]: Good catch, I meant to replace this with `pass` as child classes should implement it

File: trestle/common/const.py
Comment (enikonovad) [reply trestle/common/const.py:232]: Good suggestion, didn't think of modifying regex. Will adjust

**Pull request merged**
